compiler/rustc_hir_typeck/src/expr.rs

```diff
@@ -1402,13 +1402,31 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         expr: &'tcx hir::Expr<'tcx>,
     ) -> Ty<'tcx> {
         let element_ty = if !args.is_empty() {
+            let mut size_ct = None;
```

Comment:
This isn't a sufficient fix. The expectation may always be `None`. For example, this still ICEs:

```
struct OppOrder<const N: u8 = 3, T = u32> {
    arr: [T; N],
}

fn main() {
    let arr = [0, 0, 0];
    let _ = OppOrder::<3, u32> { arr };
}
```